File: bots/pacifist/grid.js
/**
 * Grid utility functions for AI Code Battle.
 * Toroidal distance calculations, neighbor enumeration, and BFS.
 */

function toroidalDelta(a, b, size) {
  const d = Math.abs(a - b);
  return Math.min(d, size - d);
}

function distance2(r1, c1, r2, c2, rows, cols) {
  const dr = toroidalDelta(r1, r2, rows);
  const dc = toroidalDelta(c1, c2, cols);
  return dr * dr + dc;
}

function manhattan(r1, c1, r2, c2, rows, cols) {
  return toroidalDelta(r1, r2, rows) + toroidalDelta(c1, c2, cols);
}

function moveDir(row, col, dir, rows, cols) {
  switch (dir) {
    case "N": return [(row - 1 + rows) % rows, col];
    case "E": return [row, (col + 1) % cols];
    case "S": return [(row + 1) % rows, col];
    case "W": return [row, (col - 1 + cols) % cols];
  }
}

function posKey(r, c) {
  return `${r},${c}`;
}

module.exports = { distance2, manhattan, moveDir, posKey };

File: bots/farmer/grid.go
package main

func ToroidalManhattan(a, b Position, rows, cols int) int {
	dr := abs(a.Row - b.Row)
	dc := abs(a.Col - b.Col)
	dr = min(dr, rows-dr)
	dc = min(dc, cols-dc)
	return dr + dc
}

func distance2(a, b Position, rows, cols int) int {
	dr := abs(a.Row - b.Row)
	dc := abs(a.Col - b.Col)
	dr = min(dr, rows-dr)
	dc = min(dc, cols-dc)
	return dr*dr + dc*dc
}

type cardinalStep struct {
	pos Position
	dir string
}

func cardinalSteps(p Position, rows, cols int) []cardinalStep {
	steps := []struct {
		dr, dc int
		dir    string
	}{{-1, 0, "N"}, {0, 1, "E"}, {1, 0, "S"}, {0, -1, "W"}}
	result := make([]cardinalStep, 0, 4)
	for _, s := range steps {
		result = append(result, cardinalStep{
			pos: Position{
				Row: (p.Row + s.dr + rows) % rows,
				Col: (p.Col + s.dc + cols) % cols,
			},
			dir: s.dir,
		})
	}
	return result
}

// BFS finds the shortest path from start to goal on a toroidal grid
// using 4-directional (cardinal) movement. passable returns true if a cell can be entered.
// Returns the first direction to move, or "" if unreachable.
func BFS(start, goal Position, passable func(Position) bool, rows, cols int) string {
	if start == goal {
		return ""
	}

	type node struct {
		pos Position
		dir string
	}

	visited := map[Position]bool{start: true}
	queue := []node{}

	for _, step := range cardinalSteps(start, rows, cols) {
		if step.pos == goal && passable(step.pos) {
			return step.dir
		}
		if passable(step.pos) && !visited[step.pos] {
			visited[step.pos] = true
			queue = append(queue, node{step.pos, step.dir})
		}
	}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		if cur.pos == goal {
			return cur.dir
		}

		for _, step := range cardinalSteps(cur.pos, rows, cols) {
			if !visited[step.pos] && passable(step.pos) {
				visited[step.pos] = true
				queue = append(queue, node{step.pos, cur.dir})
			}
		}
	}

	return ""
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
